### PR TITLE
[build] replaces flake8 with ruff

### PR DESCRIPTION
**Pull request by felixT2K**
This PR:

- depends on: #1180 
- replaces flake8 with ruff (same config as flake8 before)

Benefits:
- config compatible with pyproject.toml
- much faster as flake8

### CODE REVIEW

File: .github/workflows/style.yml [hunk 1]
```diff
@@ -20,11 +20,11 @@ jobs:
         with:
           python-version: ${{ matrix.python }}
           architecture: x64
-      - name: Run flake8
+      - name: Run ruff
```
Comment (frgfm) [inline .github/workflows/style.yml:23]: Thanks! Let's rename the job on line 10 as well :pray:

File: doctr/version.py
Comment (frgfm) [inline doctr/version.py:1]: Indeed, let's change this but this is for another PR (+ adding it to gitignore)

File: pyproject.toml [hunk 2]
```diff
@@ -170,6 +170,18 @@ src_paths = ["doctr", "tests", "scripts", "references", "demo", "docs", "api"]
 skip_glob = "**/__init__.py"
 known_third_party = ["tensorflow", "torch", "torchvision", "wandb", "fastprogress"]
 
+[tool.ruff]
+ignore = ["E402", "F403", "E731"]
+exclude = [".github", ".git", "venv*", "docs", "build"]
```
Comment (frgfm) [inline pyproject.toml:175]: same here (if there are python script in .github or docs, it might be nice to have those checked by ruff?)

File: setup.py [hunk 3]
```diff
@@ -9,7 +9,7 @@
 from setuptools import setup
 
 PKG_NAME = "python-doctr"
-VERSION = os.getenv("BUILD_VERSION", "0.7.1a0")
+VERSION = os.getenv("BUILD_VERSION", "0.6.1a0")
```
Comment (frgfm) [inline setup.py:12]: Thanks for that fix, but I also suggest to keep that for another PR :)

**Review by frgfm — COMMENTED** [review]
Thanks a lot @felixT2K :pray: 

Two high-level comments:
- loving the suggestion to switch to ruff, I fully agree with that. I only added a few comments to make sure we have no regression from flake8 previous config
- for the rest, everything is on point, but let's make one objective per PR and focus that one on ruff if that's ok :smiley:

File: pyproject.toml [hunk 4]
```diff
@@ -170,6 +170,18 @@ src_paths = ["doctr", "tests", "scripts", "references", "demo", "docs", "api"]
 skip_glob = "**/__init__.py"
 known_third_party = ["tensorflow", "torch", "torchvision", "wandb", "fastprogress"]
 
+[tool.ruff]
+ignore = ["E402", "F403", "E731"]
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply pyproject.toml:174]: The removed patterns seems to be invalid (.flake8 was unchecked) i will take a look again

File: doctr/version.py
Comment (felixdittrich92) [reply doctr/version.py:1]: Is already added in gitignore don't know how this could happen before 😅

File: docs/build.sh [hunk 5]
```diff
@@ -5,7 +5,14 @@ function deploy_doc(){
...
     echo "Creating doc at commit" $COMMIT "and pushing to folder $2"
-    pip install -U ..
+    # Hotfix
+    if [ "$1" \< "dcbb21f" ]
+        then
+            pip install -U ..
+            pip install rapidfuzz==2.15.1
+        else
+            pip install -U ..
+        fi
```
[the remark this answers is not in the document]
Comment (felixdittrich92) [reply docs/build.sh:15]: Will open another one for the Version and this

File: .github/workflows/style.yml (hunk 1 above)
Comment (felixdittrich92) [reply .github/workflows/style.yml:23]: Good catch i totally missed this :D

File: pyproject.toml (hunk 2 above)
Comment (felixdittrich92) [reply pyproject.toml:175]: Sure i have no use Case in mind currently but why not :)

**Comment by felixT2K** [comment]
@frgfm Do you think it would make sense to add pre-commit also ? :) 
(I personally like it one of my current colleagues hates it like hell 😂 )
I was so free to merge this minor stuff directly in #1180 think that's fine

File: pyproject.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixT2K) [reply pyproject.toml:174]: yep for the removed onces it says: Unknown rule selector

[the remark this answers is not in the document]
Comment (felixT2K) [reply pyproject.toml:174]:
```
By default, Ruff enables Flake8's E and F rules. Ruff supports all rules from the F category, and a [subset](https://beta.ruff.rs/docs/rules/#error-e) of the E category, omitting those stylistic rules made obsolete by the use of an autoformatter, like [Black](https://github.com/psf/black).
```

File: pyproject.toml [hunk 6]
```diff
@@ -70,7 +70,7 @@ testing = [
     "requests>=2.20.0",
 ]
 quality = [
-    "flake8>=3.9.0",
+    "ruff>=0.0.260",
```
Comment (frgfm) [inline pyproject.toml:73]: out of security, I would have added an upper bound as well 0.1.0 for instance to avoid troubles with the CI later but nothing major :+1:

**Review by frgfm — APPROVED** [review]
Nice, thanks for the migration Felix :+1: 
(I just noticed @felixdittrich92 & @felixT2K are the same haha)

**Comment by felixdittrich92** [comment]
> Nice, thanks for the migration Felix 👍 (I just noticed @felixdittrich92 & @felixT2K are the same haha)

Yeah it's my work profile 😅
